pkg/client/delete.go

```diff
@@ -0,0 +1,107 @@
+/*
+Copyright 2018 Heptio Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package client
+
+import (
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	kubeerror "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+)
+
+const (
+	clusterRoleFieldName  = "component"
+	clusterRoleFieldValue = "sonobuoy"
+
+	e2eNamespacePrefix = "e2e-"
+)
+
+func (c *SonobuoyClient) Delete(cfg *DeleteConfig, client kubernetes.Interface) error {
+	// Delete the namespace
+	log := logrus.WithFields(logrus.Fields{
+		"kind":      "namespace",
+		"namespace": cfg.Namespace,
+	})
+
+	err := client.CoreV1().Namespaces().Delete(cfg.Namespace, &metav1.DeleteOptions{})
+	if err := logDelete(log, err); err != nil {
+		return errors.Wrap(err, "couldn't delete namespace")
+	}
+
+	if cfg.EnableRBAC {
+		// ClusterRole and ClusterRoleBindings aren't namespaced, so delete them seperately
+		selector := metav1.AddLabelToSelector(
+			&metav1.LabelSelector{},
+			clusterRoleFieldName,
+			clusterRoleFieldValue,
+		)
+
+		deleteOpts := &metav1.DeleteOptions{}
+		listOpts := metav1.ListOptions{
+			LabelSelector: metav1.FormatLabelSelector(selector),
+		}
+
+		err := client.RbacV1().ClusterRoleBindings().DeleteCollection(deleteOpts, listOpts)
+		if err := logDelete(logrus.WithField("kind", "clusterrolebindings"), err); err != nil {
+			return errors.Wrap(err, "failed to delete cluster role binding")
+		}
+
+		// ClusterRole and ClusterRole bindings aren't namespaced, so delete them manually
+		err = client.RbacV1().ClusterRoles().DeleteCollection(deleteOpts, listOpts)
+		if err := logDelete(logrus.WithField("kind", "clusterroles"), err); err != nil {
+			return errors.Wrap(err, "failed to delete cluster role")
+		}
+	}
+
+	// Delete any dangling E2E namespaces
```

Comment:
I don't think we should do this, this is the responsibility of ginko. I plumbed through the signals we needed to make ginko clean up correctly and if it isn't, that indicates a problem somewhere. This would hide that symptom from us.

Reply:
The data would still be in the tarball, so I don't think it's hidden.  But it does raise a question of ownership, e.g. what happens if someone creates their own test jiggery based on upstream, we'd be destroying it.  Perhaps we can add a --all flag for this...?

Reply:
`sb delete --all` ~= yolo / dust off nuke it from orbit!

Reply:
So long as deleting e2e namespaces is not the default I am not picky about how this gets restructured.